Retoot add accts closes #155

alexjgriffith opened the pull request:
## Description
When responding to toots the full acct for both local and federated accounts are now added to the new toot buffer.


## Changes
- Added a function in mastodon.el to return the current user acct
- Added mastodon-toot--process-local, which takes an acct and appends the current server if it is local returns an empty string if the acct matches the current user and does only adds a prefix @ if the acct is federated
- mastodon-toot--mentions will return a formatted string of mentions or an empty string
- adds tests for mastodon-toot--mentions

## Edit (more changes)
- adds a missing , in mastodon-http--post
- `mastodon-toot--reply` now passes `mastodon-toot` a toot-id rather than the whole json
- 'mastodon-toot--reply-to-id is now a local var in a new toot
- only removes one trailing char rather than two from content in toot

hdurer commented on lisp/mastodon-toot.el:
```diff
@@ -142,14 +141,53 @@ Set `mastodon-toot--content-warning' to nil."
+         (mentions (mastodon-toot--mentions toot)))
+    (mastodon-toot (when user (concat (mastodon-toot--process-local user)
+                                      mentions))
+                   toot)))
```
That second arg is meant to be an id not the whole toot.

hdurer commented on lisp/mastodon-auth.el:
```diff
@@ -43,6 +43,9 @@
   "Alist of User access tokens keyed by instance url.")
 
+(defvar mastodon-auth--acct-alist nil
+  "Alist of acct names keyed by instance url.")
```
Is it just the account names or the whole alist of account details?

alexjgriffith replied:
Just the account names

hdurer commented on lisp/mastodon-toot.el:
```diff
@@ -142,14 +141,53 @@ Set `mastodon-toot--content-warning' to nil."
       (mastodon-http--triage response
                              (lambda () (message "Toot toot!"))))))
 
+(defun mastodon-toot--get-acct()
```
Unused function.

alexjgriffith replied:
This function is meant to be used interactively by the user. Is there a specific way to mark such a function?

hdurer commented on lisp/mastodon-toot.el:
```diff
@@ -142,14 +141,53 @@ Set `mastodon-toot--content-warning' to nil."
+
+(defun mastodon-toot--process-local (acct)
+  "Adds domain to local ACCT and replaces the curent user name with \"\"."
+  (cond ((string-match-p "@" acct) (concat "@" acct " "))
```
Can you add comments, which cases you are dealing with here?

Also: why are we adding a white space?

alexjgriffith replied:
I add the white space here to account for the nil case. In mapconcat an extra space will be added.

hdurer commented on lisp/mastodon-toot.el:
```diff
@@ -142,14 +141,53 @@ Set `mastodon-toot--content-warning' to nil."
+             nil))))
+
+(defun mastodon-toot--process-local (acct)
+  "Adds domain to local ACCT and replaces the curent user name with \"\"."
```
I don't understand this comment.

Does `acct` have to be a local name?  What is the "current user name"?

hdurer commented on lisp/mastodon-toot.el:
```diff
@@ -142,14 +141,53 @@ Set `mastodon-toot--content-warning' to nil."
+(defun mastodon-toot--process-local (acct)
+  "Adds domain to local ACCT and replaces the curent user name with \"\"."
+  (cond ((string-match-p "@" acct) (concat "@" acct " "))
+        ((string-match-p (mastodon-auth--user-acct) acct) "")
```
Ah, is this removing the current user?

That seems orthogonal to what this function is meant to do.

Remove this and explicitly remove the current user from the list you produce with this?

hdurer replied:
Also: use `string=` rather than `string-match-p`.

hdurer commented on lisp/mastodon-toot.el:
```diff
@@ -142,14 +141,53 @@ Set `mastodon-toot--content-warning' to nil."
+  (cond ((string-match-p "@" acct) (concat "@" acct " "))
+        ((string-match-p (mastodon-auth--user-acct) acct) "")
+        (t (concat "@" acct "@"
+                   (cadr(split-string mastodon-instance-url "/" t)) " "))))
```
nit: space after `cadr`.

hdurer commented on lisp/mastodon-toot.el:
```diff
@@ -142,14 +141,53 @@ Set `mastodon-toot--content-warning' to nil."
+(defun mastodon-toot--mentions (status)
+  "Extract mentions from STATUS and process them into a string."
+  (interactive)
+  (let ((mentions (cdr(assoc 'mentions status))))
```
nit: space after `cdr`

hdurer commented on lisp/mastodon-toot.el:
```diff
@@ -142,14 +141,53 @@ Set `mastodon-toot--content-warning' to nil."
+        (mapconcat (lambda(x) (mastodon-toot--process-local
+                               (cdr(assoc 'acct x))))
+                   ;; reverse does not work on vectors in 24.5
+                   (reverse (mapcar (lambda(x) x) mentions)) "") "")))
```
- There is a function `identity` which could replace your lambda
- The [manual](https://www.gnu.org/software/emacs/manual/html_node/elisp/Vector-Functions.html) mentions `(append avector nil)` as a way to convert a vector to a list.

alexjgriffith replied:
Perfect, was looking for a way to do the conversion cleanly, this seemed unnecessary.

hdurer commented on lisp/mastodon-toot.el:
```diff
@@ -142,14 +141,53 @@ Set `mastodon-toot--content-warning' to nil."
+        (mapconcat (lambda(x) (mastodon-toot--process-local
+                               (cdr(assoc 'acct x))))
+                   ;; reverse does not work on vectors in 24.5
+                   (reverse (mapcar (lambda(x) x) mentions)) "") "")))
```
Put that third arg to `mapconcat` on its own line to make it clearer.

Also: that should probably be a space `" "` so that names are properly separated and you don't have to add the space in `mastodon-toot--process-local`.

alexjgriffith replied:
I add the spaces in `mastodon-toot-process local` to save us having to remove repeated spaces introduced from the null case (where your username is in the reply list).

hdurer commented on test/mastodon-toot-tests.el:
```diff
@@ -1,6 +1,46 @@
+  '((mentions . [])))
+
+(ert-deftest toot-multi-mentions ()
+  (should (string= (let ((mastodon-auth--acct-alist '(("https://local.social". "null")))
```
I think this is more readable if you pull the `let` out.
That's basically the setup of the test context and now part of the actual test.

hdurer commented on lisp/mastodon-auth.el:
```diff
@@ -79,5 +82,20 @@ Generate token and set if none known yet."
     token))
 
+(defun mastodon-auth--get-account ()
+  "Request user credentials and return an acct."
```
Nit "account name"  Maybe even make that clear in the function name?

hdurer commented on lisp/mastodon-tl.el:
```diff
@@ -318,7 +318,7 @@ also render the html"
     (mastodon-tl--spoiler toot)
     ;; remove two trailing newlines
-    (substring (mastodon-tl--content toot) 0 -2)
+    (substring (mastodon-tl--content toot) 0 -1)
```
That comment above talks about 2 spaces being removed...

hdurer commented on lisp/mastodon-toot.el:
```diff
@@ -142,14 +141,61 @@ Set `mastodon-toot--content-warning' to nil."
+eg. \"yourusername\" -> \"\"
+eg. \"feduser@fed.social\" -> \"feduser@fed.social\""
+  (cond ((string-match-p "@" acct) (concat "@" acct " ")) ; federated acct
+        ((string-match-p (mastodon-auth--user-acct) acct) "") ; your acct
```
Again: why not `string=` ?  Is there extra whitespace somewhere to be ignored?

hdurer commented on lisp/mastodon-toot.el:
```diff
@@ -142,14 +141,61 @@ Set `mastodon-toot--content-warning' to nil."
+  "Extract mentions from STATUS and process them into a string."
+  (interactive)
+  (let ((mentions (cdr (assoc 'mentions status))))
+    (if (> (length mentions) 0)
```
Nit: This check isn't needed, right?  If mentions is the empty vector, the mapappend will still do the right thing.

hdurer commented on test/mastodon-toot-tests.el:
```diff
@@ -1,6 +1,47 @@
   (with-mock
     (mock (kill-buffer-and-window))
     (should (eq nil (mastodon-toot--cancel)))))
+
```
Nit: remove the empty lines at end?

hdurer reviewed (commented):
Great.  Only a few nitty comments.

alexjgriffith committed:
Retoot add accts closes #155

When responding to toots the full acct for both local and federated accounts are now added to the new toot buffer.

Changes
- Added a function in mastodon.el to return the current user acct
- Added mastodon-toot--process-local, which takes an acct and appends the current server if it is local returns an empty string if the acct matches the current user and does only adds a prefix @ if the acct is federated
- mastodon-toot--mentions will return a formatted string of mentions or an empty string
- adds tests for mastodon-toot--mentions
- adds a missing , in mastodon-http--post
- `mastodon-toot--reply` now passes `mastodon-toot` a toot-id rather than the whole json
- 'mastodon-toot--reply-to-id is now a local var in a new toot
